lib/db/leveldb.go

```diff
@@ -166,27 +175,57 @@ func globalKeyFolder(key []byte) []byte {
 	return folder[:izero]
 }
 
-type deletionHandler func(db dbReader, batch dbWriter, folder, device, name []byte, dbi iterator.Iterator) int64
+type deletionHandler func(t dbRWTransaction, folder, device, name []byte, dbi iterator.Iterator) int64
+
+func (db *dbInstance) newReadOnlyTransaction() dbROTransaction {
+	snap, err := db.GetSnapshot()
+	if err != nil {
+		panic(err)
+	}
+	return dbROTransaction{
+		Snapshot: snap,
+		db:       db.DB,
+	}
+}
+
+func (db *dbInstance) newReadWriteTransaction() dbRWTransaction {
+	t := db.newReadOnlyTransaction()
+	return dbRWTransaction{
+		dbROTransaction: t,
+		Batch:           new(leveldb.Batch),
+	}
+}
+
+func (t dbROTransaction) close() {
+	t.Release()
+}
+
+func (t dbRWTransaction) close() {
```

Comment:
Potentially this should be called commit() instead, but that would mean that the code

```
t := db.newReadWriteTransaction()
defer t.close()
```

is completely valid and does something entirely different that what you might expect, for example if you refactor something to need an rw-transaction after having just an ro-transaction... So that's why it is what it is. At first I had commit() on both types, but it seemed to make even less sense to commit() a read only transaction...